File: DSA-C++/6-array/17-majority-element.cpp
#include <bits/stdc++.h>

using namespace std;
// Given an array nums of size n, return the majority element. The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.


// brute force - O(n²), O(1)
int majority_element_1(vector <int> &nums) {
    int n = nums.size(), max_cnt=0, maj_ele=nums[0];

    for (int i=0; i<n; ++i) {
        int cnt = 0;

        for (int j=0; j<n; ++j) {
            if (nums[i] == nums[j]) {
                cnt++;
            }
        }

        if (cnt > max_cnt) {
            max_cnt = cnt;
            maj_ele = nums[i];
        }
    }
    
    return max_cnt > n/2 ? maj_ele : -1;
}


// hashing - O(n), O(n)
int majority_element_2(vector <int> &nums) {
    int n = nums.size(), max_freq=0, maj_ele=nums[0];
    unordered_map <int,int> umap;

    for (int &x:nums) {
        umap[x]++;
    }

    for (auto [k,v]:umap) {
        if (v > max_freq) {
            max_freq = v;
            maj_ele = k;
        }
    }

    return max_freq > n/2 ? maj_ele : -1;
}


// Boyer-Moore Majority Vote Algorithm - O(n), O(1)
int majority_element_3(vector <int> nums) {
    int n = nums.size(), cnt=0, ele=nums[0];

    for (int i=0; i<n; ++i) {
        if (cnt == 0) {
            ele = nums[i];
            cnt++;
        }
        else if (ele != nums[i]) {
            cnt--;
        }
        else {
            cnt++;
        }
    }


    // verify
    cnt = 0;
    for (int &x:nums) {
        if (x == ele) {
            cnt++;
        }
    }

    return cnt > n/2 ? ele : -1;
}

int main() {
    vector <int> nums = {2, 2, 1, 1, 1, 2, 2};
    nums = {7, 7, 7, 7, 7};
    nums = {1, 2, 1, 3, 1, 4, 1};
    nums = {1, 2, 3, 4, 9, 9, 9, 9, 9};
    // nums = {1, 2, 3};       // edge case: majority element does not exists

    
    cout << majority_element_1(nums) << endl;
    
    cout << majority_element_2(nums) << endl;

    cout << majority_element_3(nums) << endl;


    return 0;
}

File: DSA-Python/5-array/17-majority-element.py
# Given an array nums of size n, return the majority element.
# The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.



# brute force - O(n²), O(1)
def majority_element_1(nums: list[int]) -> int:
    n, major, max_cnt = len(nums), 0, 0
    
    for i in range(n):
        cnt = 0
        
        for j in range(i, n):
            if nums[i] == nums[j]:
                cnt += 1
                
        if cnt > max_cnt:
            max_cnt = cnt
            major = nums[i]
            
    return major if max_cnt > n//2 else -1



# hashing - O(n), O(n)
def majority_element_2(nums:list[int]) -> int:
    n, major = len(nums), 0
    
    umap = {}
    
    for x in nums:
        umap[x] = umap.get(x, 0) + 1
        
    for (k,v) in umap.items():
        if v > n//2:
            return k
        
    return -1



# booyer moores voting algorithm - O(n), O(1)
def majority_element_3(nums:list[int]) -> int:
    n, cnt, major = len(nums), 0, 0
    
    for i in range(n):
        if cnt == 0:
            major = nums[i]
            cnt = 1
        elif major == nums[i]:
            cnt += 1
        else:
            cnt -= 1
            
    cnt = nums.count(major)
    
    return major if cnt > n//2 else -1
        
        
    

if __name__ == "__main__":
    loop = int(input())

    for _ in range(loop):
        nums = list(map(int, input().split(' ')))
        
        res = majority_element_1(nums)
        
        res = majority_element_2(nums)
        
        res = majority_element_3(nums)
        
        print(res)
        
        
        
'''
5
2 1 2 3 2
7
1 2 3 4 4 4 4
-1 -1 -1 2 3
1 2 1 2 1 2 2
'''